### DISCUSSION TITLE
SISAP Challenge 2023 Ranking

### DISCUSSION THREAD

**Discussion by sadit** [sisap-challenges/challenge2023#16] [Announcements]
# Rankings for the SISAP 2023 Challenge

Dear team members @cole-foster @dohnal @TerkaSlan @eth42 @flopper123 @VladimirMic @matsui528

We are glad to share a summary of the SISAP challenge results. In particular, the ranking per subset per track; more detailed results will be listed in the overview and discussed during the challenge track of the SISAP 2023 conference. Please recall that Task descriptions are detailed on the site of the SISAP challenge <https://sisap-challenges.github.io/tasks/>.

We prepared docker Linux images using the submitted solutions using a 28-core (56 threads) Intel(R) Xeon(R) CPU E5-2690 V4 workstation with 512GiB of RAM. We encourage participants to use multithreading or multiprocessing in the construction and searching stages to take advantage of the hardware.

We included different baselines to compare with previous work. Note that these are far from trivial solutions.

- BASELINE-SearchGraph: Uses the `SearchGraph` index from the Julia's package [`SimilaritySearch.jl`](https://github.com/sadit/SimilaritySearch.jl). See the _Julia example_ at the `validation` branch <https://github.com/sisap-challenges/sisap23-laion-challenge-julia-example/tree/validation>.
- BASELINE-FAISS-HNSW: Uses the HNSW index from faiss [`faiss`](https://faiss.ai/). See the _Python faiss_ at the `validation` branch <https://github.com/sisap-challenges/sisap23-laion-challenge-faiss-example/tree/validation>.
- Bruteforce: Uses exhaustive search from Julia's package `SimilaritySearch.jl`.

Thank you for your participation, 
@maumueller @zevahcle @sadit 

# TASK A

Uses the 768-dimensional LAION vectors (computed with a CLIP model). We use 10k private queries; $k=10$.
The lists only show the ranking of solutions exceeding 90% recall (best by team). 

## 10 Million

| Team                                  | Size | Rank | 
| ---                                      | ---- |-----:|
| HSP                                    | 10M  | 1    |
| UTokyo                               | 10M  | 2    |
| BASELINE-SearchGraph  | 10M  | 3    |
| BASELINE-FAISS-HNSW  | 10M  | 4    |
| HIOB                 | 10M  | 5    |
| CRANBERRY            | 10M  | 6    |
| LMI                  | 10M  | 7    |
| Bruteforce           | 10M  | 8    |
  
## 30 Million

| Team               | Size | Rank | 
| ---                | ---- |-----:|
HSP                  | 30M  |  1
UTokyo               | 30M  |  2
BASELINE-FAISS-HNSW  | 30M  |  3
BASELINE-SearchGraph | 30M  |  4
HIOB                 | 30M  |  5
CRANBERRY            | 30M  |  6
Bruteforce           | 30M  |  7

## 100 Million

| Team               | Size | Rank | 
| ---                | ---- |-----:|
HSP                  | 100M | 1
BASELINE-SearchGraph | 100M | 2
BASELINE-FAISS-HNSW  | 100M | 3
HIOB                 | 100M | 4
CRANBERRY            | 100M | 5
UTokyo               | 100M | OOM
Bruteforce           | 100M | NR

- OOM: Out of memory
- NR: Not run

# TASK B
This task uses the LAION dataset encoded as a bit vector. It utilizes 10,000 private queries. We compute $k$nn queries using $k=10$.

- BASELINE: a 1024-bit projection using the encoding described in _brief permutations_ [@SCT14].
- The search was evaluated with the exhaustive search of Julia's package `SimilaritySearch.jl`.

[@SCT14]: _Santoyo, F., Chávez, E., & Téllez, E. S. (2014). A compressed index for hamming distances. In Similarity Search and Applications: 7th International Conference, SISAP 2014, Los Cabos, Mexico, October 29-31, 2014. Proceedings 7 (pp. 113-126). Springer International Publishing._

## 10 Million

| Team     | #bits | Recall 
| ---      | ---   | ---:|
| HIOB     | 1024  | 1
| HIOB     | 256   | 2
| BASELINE | 1024  | 3

## 30 Million

| Team     | #bits | Recall 
| ---      | ---   | ---:|
| HIOB     | 1024  | 1
| HIOB     | 192   | 2
| BASELINE | 1024  | 3

## 100 Million

| Team     | #bits | Recall 
| ---      | ---   | ---:|
| HIOB     | 1024  | 1
| BASELINE | 1024  | 2
| HIOB     | 128   | 3

# TASK C
This task asks for solutions for fast indexing binary sketches under the Hamming distance. Participants can use their Task B approach to compute the projections or our provided bit-vectors. Note that reranking with original CLIP vectors is invalid under this task.

- Indexing binary hamming vectors. Bruteforce recall is around 0.24; we show the best-performing parameters exceeding recall at 0.22 (minimum recall).

## 10 Million

| Team | Size | Rank | 
| ---  | ---- |-----:|
BASELINE-SearchGraph | 10M | 1
HIOB | 10M | 2
Bruteforce | 10M | 3
SWANN | 10M |  4

## 30 Million

| Team | Size | Rank | 
| ---  | ---- |-----:|
BASELINE-SearchGraph | 30M | 1
HIOB | 30M | 2
Bruteforce | 30M | 3
SWANN | 30M |  4

## 100 Million

| Team | Size | Rank | 
| ---  | ---- |-----:|
BASELINE-SearchGraph | 100M | 1
Bruteforce | 100M | 2
SWANN | 100M |  3
HIOB | 100M | UMT

UMT: Below the minimum threshold

**Comment by dohnal** [comment]
Thank you for advertising the ranking. Is there a possibility to see the details (recall, query evaluation time, index construction time, memory requirements)?

**Reply by maumueller** [reply]
An overview with more detailed results can be found at https://github.com/sisap-challenges/challenge2023/blob/main/results/RESULTS.md, with plots for [10M](https://github.com/sisap-challenges/challenge2023/blob/main/results/result_10M.png), [30M](https://github.com/sisap-challenges/challenge2023/blob/main/results/result_30M.png), [100M](https://github.com/sisap-challenges/challenge2023/blob/main/results/result_100M.png). Raw results are visible here: https://github.com/sisap-challenges/challenge2023/blob/main/results/res.csv

**Comment by kno10** [comment]
Task C: "Note that reranking with original CLIP vectors is invalid under this task."
I did not see that in the original task description. Is that why HIOB is not included? It is using hamming distances.
Setting the number of candidates to exactly k will disable refinement, so there are no code changes required to satisfy this requirement.

**Reply by sadit** [reply]
Hi @kno10,


I will open an issue and mention you on our fork (of your submission) to handle this.

Best,
Eric

**Comment by flopper123** [comment]
Dear SISAP team,

I think we might have misunderstood the result format.
Our query time in the results were given as the total number of seconds it took to run all queries. This is the same as shown in the example using FAISS (https://github.com/sisap-challenges/sisap23-laion-challenge-faiss-example/blob/master/search/search.py).
However, now that we look at the result, it seems to be noted as seconds per query, which causes our solution to be 10_000 times slower than it's actual performance. Can we validate, that this is the case?

The logging of time happens [in this file](https://github.com/flopper123/swann/blob/feature/sisap/src/sisap/sisap.cpp) on lines 83 - 92.

Best regards,
SWANN Team

**Comment by sadit** [comment]
Dear Thomas,


We are using the accumulated query time for 10k objects for the ranking.

I checked the lines you mention, total_time seems the one we should use.

You can check all submissions on the SISAP Challenges organization.

https://github.com/orgs/sisap-challenges/repositories￼


For the baselines:
- https://github.com/sisap-challenges/sisap23-laion-challenge-julia-example/blob/ea4d81b751dd00cd81579a6cd5a1d05fb94a3341/src/run.jl#L97
- https://github.com/sisap-challenges/sisap23-laion-challenge-faiss-example/blob/2e4dc6231872e48e1b5732f383c95f1a790b4b64/search/search.py#L104


Best,
Eric

**Comment by eth42** [comment]
The sketch lengths for HIOB Task B seem to be wrong in the table above. I was curious about the drop in recall in the [extended ranking table](https://github.com/sisap-challenges/challenge2023/blob/main/results/RESULTS.md), which did not happen in our experiments, and cross-referenced with the [reported result table](https://github.com/sisap-challenges/challenge2023/blob/main/results/res.csv). In there, the sketch lengths for 30M are given with 192 bits and for 100M with 128 bits. That would agree with our experimental results and also explain the drop in recall. In our experiments, the recall for a constant sketch length did not change significantly (in fact it slightly increased but that might not be reliable).

**Reply by sadit** [reply]
Hi @eth42 

It is possible. However, entries in the CSV file are filled directly from your h5-file result file; we just compute the recall.

It is also possible that the differences are produced because we used another query set. I will prepare private queries and its gold-standard companion so people can use them to verify or, I hope, in future publications. I will be doing this today.

Best,
Eric

**Reply by eth42** [reply]
I am specifically referring to the lines 738/739, 741/744, and 747/748 (pairwise almost identical lines, hence the X/Y notation) which seem to be the source for the recall values used for the ranking. At least the recall values are the same and these are the only lines that relate to HIOB and have exactly these recall values. The `parameters` column references the number of bits as 128, 192, and 256 (`n_bits=[...]`). So I assume the evaluation is based on runs with these numbers of bits per sketch, which is 256 for 10M, 192 for 30M and 128 for 100M.

Edit: For Task B we primarily intended the 1024 bit version, so I assume you used the h5 files resulting from the Task A/C workflows, which had differently sized sketches for a better trade-off of Hamming and Euclidean distance costs.

**Reply by sadit** [reply]
For task B we used https://github.com/sisap-challenges/sisap2023-challenge-submission-HIOB/blob/main/sisap23-run-taskB.sh that is a copy of GHA workflows with the prefix TaskB_

**Reply by eth42** [reply]
These are correct but only the 1024-bit runs. The table above contains recall values for smaller HIOB sketches created with 128/192/256 bits, that are all annotated as 256-bit in the table, where I think it should be 128/192/256. It would explain why the rank switches from 30M to 100M for the "smaller" HIOB and the baseline, since the number of bits decreases from 192 at 30M to 128 at 100M.
Results are all correct, just the manually written table above is wrong, if I see that correctly. And if so, same goes for the [extended table](https://github.com/sisap-challenges/challenge2023/blob/main/results/RESULTS.md).

**Reply by sadit** [reply]
I understand now; thank you for noticing the issue. I put task B's descriptions; sorry for the mistake. I will update the table. Note that we consider only the best setup per team (1024-bit), and we will remove extra entries in the overview.

Best,
Eric

**Comment by sadit** [comment]
Hi @cole-foster @dohnal @TerkaSlan @eth42 @flopper123 @VladimirMic @matsui528


We released the private query set and gold standards for the 10, 30, and 100M subsets on the official SISAP Indexing Challenge site:

https://sisap-challenges.github.io/datasets/

It follows the same storage conventions as the public dataset. If you find these data useful, please cite the oncoming overview paper instead of the sisap challenge site.


Thank you,
@maumueller @zevahcle @sadit
